sony.go

```diff
@@ -0,0 +1,224 @@
+// is returned.
+func NewSonyError(resp []any) error {
+	if len(resp) != 2 {
+		return ErrInvalidSonyResponse
+	}
+	code, ok := resp[0].(float64) // float64 as JSON decodes numbers as float64
+	if !ok {
+		return ErrInvalidSonyResponse
+	}
+	msg, ok := resp[1].(string)
+	if !ok {
+		return ErrInvalidSonyResponse
+	}
+	return SonyError{Code: int(code), Message: msg}
+}
+
+// Error returns the message portion of the error.
+func (err SonyError) Error() string {
+	return err.Message
+}
+
+// Unwrap returns ErrSony for all errors so SonyErrors can easily be checked
+// with `errors.Is(err, ErrSony)` for when you don't care about the specific
+// code or message of the error (when you would use `errors.As()`).
+func (err SonyError) Unwrap() error {
+	return ErrSony
+}
+
+// NewRESTClient creates and returns a BraviaClient reachable at the given
+// hostname, using the Pre-Shared Key given as psk as the password. If psk is
+// the empty string, it is not used.
+func NewRESTClient(hostname, psk string) *RESTClient {
+	return &RESTClient{
+		BaseURL: "http://" + hostname + "/sony",
+		PSK:     psk,
+		HTTPClient: &http.Client{
+			// Timeout after 10s. Arguably that's too long.
+			// This doesn't really need to be configurable.
+			Timeout: 10 * time.Second,
+		},
+	}
+}
+
+// empty is a type to be used with `post[T]()` for when a response is not returned.
+// e.g. `_, err := post[empty](...)`.
+type empty struct{}
+
+// post[T] executes a REST IP control command returning the result of type T or
+// an error if the command did not succeed. If no data was returned from the
+// HTTP call, the returned value will be nil. The `empty` type can be used when
+// no response is expected:
+//
+//	_, err := post[empty](client, service, method, version, params...)
+//
+// The protocol docs define service, method and version. Params are any values
+// that can be marshaled as JSON and are passed in the `params` part of the
+// request.
+//
+// The `result` field in the JSON response will be unmarshaled into a variable
+// of type T and returned.
+func post[T any](c *RESTClient, service, method, version string, params ...any) (*T, error) {
+	brq, err := c.newRequest(service, method, version, params...)
+	if err != nil {
+		return nil, fmt.Errorf("new request: %w", err)
+	}
+	resp, err := c.do(brq)
+	if err != nil {
+		return nil, fmt.Errorf("http: %w", err)
+	}
+	bresp, err := decodeResp[T](resp)
+	if err != nil {
+		return nil, fmt.Errorf("decode: %w", err)
+	}
+	if len(bresp) == 0 {
+		return nil, nil //nolint:nilnil // T can be `empty` for no result expected. not an error.
+	}
+	return &bresp[0], nil
+}
+
+type restRequest struct {
+	Method  string `json:"method"`
+	Version string `json:"version"`
+	ID      int    `json:"id"`
+	Params  []any  `json:"params"`
+}
+
+func (c *RESTClient) newRequest(service, method, version string, params ...any) (*http.Request, error) {
+	if params == nil {
+		params = []any{}
+	}
+	breq := restRequest{
```

Comment:
I'll rename `breq` to `payload`. I've gotten rid of the `restRequest` struct, opting for an anonymous struct as this is the only place it is used. I want the parameters as this function is meant to make it easy to make a call - i kind of like `post[empty](c, "system", "setPowerStatus", "1.0", param)` for example. It would be multiple more lines if I had to initialise a struct.